src/crimson/os/seastore/btree/fixed_kv_btree.h

```diff
@@ -511,6 +511,9 @@ class FixedKVBtree {
           &child_node);
       } else {
         if (i->get_val().pladdr.is_laddr()) {
+          assert((!std::is_base_of_v<typename internal_node_t::base_t, child_node_t>));
+          assert(!node->children[i->get_offset()] ||
+                  is_reserved_ptr(node->children[i->get_offset()]));
```

Comment:
This should mean the indirect mapping contains ZERO (P_ADDR_ZERO).

But P_ADDR_ZERO is of type paddr_t, not laddr, right? (see L513)

So why is `is_reserved_ptr(node->children[i->get_offset()])` possible to be true?

Reply:
I think `preserved_ptr` is added when inserting to the lba tree without a corresponding extent, which happens when reserving lba regions(P_ADDR_ZERO) and cloning mappings (inserting indirect mappings).

Reply:
> I'm wondering if it is simpler not to convert the reserved region into a clone, it should be identical to create as direct mapping with `P_ADDR_ZERO` during cloning, which saves an indirection/refcounter and a corner case.

Sorry, I'm not following. I think reserved regions are not treated as clones.

Reply:
> Sorry, I'm not following. I think reserved regions are not treated as clones.

Probably I'm missing something as well.

> I think preserved_ptr is added when inserting mappings to the lba tree without corresponding extents, which happens when reserving lba regions(P_ADDR_ZERO) *or* cloning mappings (inserting indirect mappings).

Do you mean `is_reserved_ptr(node->children[i->get_offset()])` can be true here for an indirect mapping is because it is cloning mappings, not because it is finally pointing to a reserved region?

If that's correct, then I'm not following the comment from https://github.com/ceph/ceph/blob/9a6d6b4b0534e956a0eb64c3f71240536e72b666/src/crimson/os/seastore/btree/fixed_kv_node.h#L133-L139
wrotes that assigning `get_reserved_ptr()` is only possible when reserving lba spaces (i.e. reserving ZERO, right?).

Probably the comment is wrong? Or can you help to elaborate?

Reply:
> > Sorry, I'm not following. I think reserved regions are not treated as clones.
> 
> Probably I'm missing something as well.
> 
> > I think preserved_ptr is added when inserting mappings to the lba tree without corresponding extents, which happens when reserving lba regions(P_ADDR_ZERO) _or_ cloning mappings (inserting indirect mappings).
> 
> Do you mean `is_reserved_ptr(node->children[i->get_offset()])` can be true here for an indirect mapping is because it is cloning mappings, not because it is finally pointing to a reserved region?

Yes.

> 
> If that's correct, then I'm not following the comment from
> 
> https://github.com/ceph/ceph/blob/9a6d6b4b0534e956a0eb64c3f71240536e72b666/src/crimson/os/seastore/btree/fixed_kv_node.h#L133-L139
> 
> 
> wrotes that assigning `get_reserved_ptr()` is only possible when reserving lba spaces (i.e. reserving ZERO, right?).
> Probably the comment is wrong? Or can you help to elaborate?

Ah, that comment was there before the implementation of OP_CLONE, and should be corrected. Sorry for the inconvenience.

Reply:
Just corrected the comment.